### ISSUE TITLE
What are the current pain points?

### ISSUE THREAD

**Issue by rsangole** [rsangole/tidyrabbit#1]
Let's identify the current pain points of my code.

**Comment by RMHogervorst** [comment]
Let me work through some data sets and analysis plans to get a sense of where the pain points are.

**Comment by RMHogervorst** [comment]
repeated actions I notice, in your blogpost:

- apply transformation on source df, take resulting data frame and modify the names of the variables (by adding a prefix or suffix, saving the colnames in a separate character vector 

I think it would be more elegant if we combine all these steps with the enframing step

- You first combine all the transformations of the source df into 1 dataframe, then you use the `filtered columns function` to extract the columns; in a way you are re-creating the 3 dataframes you created at the start. 

- Since you apply the same preprocessing (center and scale) on all models (in this case) it might be wiser to do that first on the data before putting it into the starter data frame [but it did make me think about the cases where you do not want to do it to all variants see #2 ]


==

Suggestions:  a function, or set of functions to take the starting dataframe, apply transformations (perhaps renaming them in the process) and nesting them in a the starter dataframe

step 1, STARTERDATAFRAME: transformations (take a source dataframe, apply the transformations that you want to try out, put it all into 1 big nested frame with columns 'transformation.name', 'train.X',  'train.Y'

step 2, MODELSDATAFRAME: with columns 'modelName'  'model'

step 3, TRAIN DATAFRAME: I actually wonder if it is a good idea to repeat the data here, it might be better if we kept the data and just refer to it? [duplicating large datasets seems unwise]
Names:   'transformation.name' , 'train.X.name' ,    'train.Y'[maybe leave out too?],  'model.name'
Another way to avoid copying the dataframe is if we made it like this
`'transformation.name' , 'train.X.name' ,    'train.Y', model1, model2, model 3 , etc`
and stored the result in those cells, so that every  row contains a transformation from STARTERFRAME and the columns after train.Y contain the modelnames from MODELSDATAFRAME. 
But we could use the elegant purr solution you created:
```
train_df %<>%
    mutate(params = map2(train.X, train.Y,  ~ list(X = .x, Y = .y)),
           modelFits=invoke_map(model,params)
    )
```

I am a very big fan of the endresult of the whole thing:
```
# A tibble: 6 x 8
     id   pred_varsets  modelName  model     RMSE    RMSESD       Rsq             bestTune
  <int>          <chr>      <chr> <list>    <dbl>     <dbl>     <dbl>               <list>
1     1 preds.boxcoxed rpartModel  <fun> 5.854134 1.1701960 0.7532030 <data.frame [1 x 1]>
2     2 preds.original rpartModel  <fun> 6.035083 1.2358751 0.7175567 <data.frame [1 x 1]>
3     3       preds.sq rpartModel  <fun> 5.863157 0.7634966 0.7354118 <data.frame [1 x 1]>
4     4 preds.boxcoxed   xgbModel  <fun> 3.535163 0.5783796 0.8940754 <data.frame [1 x 7]>
```

It is exactly what you want to know at the end. 



Please note that this is all open to discussion, We might have excellent reasons to NOT simplify the repeating actions or creation of artefacts.  Also, I have the benefit of seeing your worked out code, so I don't have to think up how to make it run, only how to make it run faster or more elegant

**Comment by RMHogervorst** [comment]
If we use some `tidy eval` magic in the TRAIN DATAFRAME we could refer to the data by name only?

A function that takes the starter dataframe and the name from the train dataframe to return the correct dataset. maybe something like this. 
```
train_df %<>%
    mutate(params = map2(  training_data(df_name = starter_df, dataframe = train.X),  train.Y,  ~ list(X = .x, Y = .y)),
           modelFits=invoke_map(model,params)
    )
```
Although this would work because map2 cannot iterate over that...

**Comment by rsangole** [comment]
Lots of good ideas here and I'm going to have to think over it before I respond.

I wanted to drop a note asking if you've heard or used Kuhn's new package `recipe` at [link](https://github.com/topepo/recipes).

I've been playing around just a bit with it... and I think it would be a handy one to leverage to "automate lots of transformations/changes".

My gut feeling is this:

1. For a really thin wrapper, we should enable a quick way to get a starter DF up & running, follow the workflow I had; just minimize repeat code.
2. A little more sophistication can incorporate your ideas above + packages like `recipe` to allow the user to automate things quicker.

**Comment by RMHogervorst** [comment]
good point keeping it small and lean! Things I tend to forget while working. I'l look into recipe

**Comment by RMHogervorst** [comment]
I think we need 3 dataframes

1. Starterframe

Contains everything related to start data: [name of the transformation, preprocess object,dataframes]

2. Modelsframe

Everything related to the models: [name of model, modelobject, possibly parameters of model]

3. Resultsframe

Everything related to results
[name of transformation, name of model, results object, possibly extracted fit measures, in following steps]

**Comment by rsangole** [comment]
Yep, this makes sense. We should have a function that sets these frames up, especially 1 and 2. 
3 would be a result of actually solving 2, plus selecting what kind of post-processing options a user might select (possible fit measures etc)

**Comment by RMHogervorst** [comment]
(I will keep posting things that I notice while trying out  tidyrabbit on different dataset, they are in no way criticisms of you)


I keep copying this part in new setups

```
combined_frame %<>%
    mutate(params = map2(train.x,y ,  ~ list(X = .x, Y = .y)),
           modelFits=invoke_map(model,params)
    )
```
It's just not pretty. Also we have the trainingset and Y now in the same df twice.
